### DISCUSSION TITLE
Does the AT protocol support content signing?

### DISCUSSION THREAD

**Discussion by jmdugan** [bluesky-social/atproto#1019] [Clarifications and "Why?"] [answered]
does the AT protocol support content signing? eg, will a user be able to create a digital signature for a piece of content that can make a strong proof of provenance, regardless of(independent of) where the content is hosted/posted?

this, to me, seems like the increasingly needed next step in a world where soon *anything can be faked* by llms and ai agents. alternately, if this is not in the AT protocol, do people have ideas to either expand of extend or interact with AT to support this use case?

thank you

**Comment by snarfed** [comment]
https://atproto.com/guides/overview#account-portability

**Comment by bnewbold** [comment, marked as answer]
Content in an account's repository is always signed, using the "atproto signing key" from the account's DID document. This key can change over time (key rotation), but the most recent version of the repo (the most recent commit) should always be signed by the currently valid signing key.

This design, which is similar to signed commits/tags in git, is a classic use-case for Merkle trees, and was intentionally chosen over signing each piece of content individually. Signing each piece of content is much more resource intensive (of CPU, storage, and bandwidth), and makes key rotation much more expensive.

Content outside of repositories, such as labels, may also be signed in the future. Each use of cryptography requires care and thought, so we are going slow and don't have a generic signing mechanism, and intentionally do not support encryption/decryption using the atproto signing key for now.
